### PR TITLE
Add "visualStart" to Fix Effect Placement Bug

### PR DESCRIPTION
**Pull request by Ferociousfeind**
**Bugfix:** This PR addresses the long-standing issue (in #5857 and lesser-so mentioned elsewhere) that many effects are placed incorrectly at high velocities. (Well, at any velocity other than 0, but the error is proportional with velocity)

## Fix Details
Adds a sister Point variable, "visualStart", which mirrors "start", without the -0.5x velocity compensation. This fixes the issue!
I do not think this change has _any_ wider consequences other than fixing this visual bug. Might as well push this to master, no?

## Testing Done
image
image
(Mind the plugin content, but this is before and after the change. Notice the streaky stars in the background, we're moving very quickly here.)

## Save File
[bee person.txt](https://github.com/endless-sky/endless-sky/files/8544746/bee.person.txt)
Use this save file, take off from Trinket in the modified (low-drag) Emerald Sword, pick up speed and fire the Dragonflame Cannon to observe whether or not the _effect_ which replaces a projectile is placed correctly. This is my personal all-content save file for testing things, so it might be a little bloated. This save file will work without any plugins- the changes to the ES are in the save file, as all owned ship data is.

### CODE REVIEW

**Comment by DeBlister** [comment]
I'm not really sure what the benefit to the -0.5x velocity is.  What makes ES projectiles different than projectiles in the real world, where they inherit 100% of their parent's velocity (barring air resistance/drag, which I believe is accounted for in its own way)?  I'm in favor of the visual workaround certainly, but should this be a workaround rather than a fix?

**Comment by Ferociousfeind** [comment]
The "start" Point is used to correctly place projectiles but it was incorrectly also used to place effects. (The velocity inclusion is a positional offset, and doesn't impact actual velocity velocity of anything. Purely positional fiddlings.)

I would not be so sure it's a workaround, I think applying "visualStart" in other places, where "start" is used now, would introduce the opposite bug in projectiles- appearing ahead of their parent ship instead of perfectly on top.

File: source/Hardpoint.cpp [hunk 1]
```diff
@@ -218,6 +218,7 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
 	// offset of (.5 * velocity) and that velocity includes the velocity of the
 	// ship that fired them.
```
Comment (quyykk) [inline source/Hardpoint.cpp:219]: This comment isn't accurate anymore.

File: source/Hardpoint.cpp [hunk 2]
```diff
@@ -227,7 +228,7 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
 	projectiles.emplace_back(ship, start, aim, outfit);
 
 	// Create any effects this weapon creates when it is fired.
-	CreateEffects(outfit->FireEffects(), start, ship.Velocity(), aim, visuals);
+	CreateEffects(outfit->FireEffects(), visualStart, ship.Velocity(), aim, visuals);
```
Comment (quyykk) [inline source/Hardpoint.cpp:231]: If we create projectile effects at a different position than they actually are wouldn't that mean that we might see a visual bug where a very fast projectile hits something while its effect lags behind and appears as if it hasn't hit the ship?

Comment (DeBlister) [reply source/Hardpoint.cpp:231]: This concerns me as well, but Ferocious pointed out that maybe normal projectiles need the 0.5x velocity.  I'll have to look at what normal projectiles do without it in addition to what the Dragonflame Cannon does.

File: source/Hardpoint.cpp (hunk 1 above)
Comment (Amazinite) [reply source/Hardpoint.cpp:219]: Well projectiles still are using an offset of half the velocity as the comment mentions. It's the effects that aren't using the offset anymore, but the comment doesn't mention them. (Although we could probably update this comment to describe *why* this offset is needed to "Get projectiles to start at the right position.")

File: source/Hardpoint.cpp [hunk 3]
```diff
@@ -218,6 +218,7 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
 	// offset of (.5 * velocity) and that velocity includes the velocity of the
 	// ship that fired them.
 	Point start = ship.Position() + aim.Rotate(point) - .5 * ship.Velocity();
+	Point visualStart = ship.Position() + aim.Rotate(point);
```
Comment (Amazinite) [inline source/Hardpoint.cpp:221]: Notably, `start` later has `aim.Rotate(outfit->HardpointOffset())` added to it. `visualStart` now does not. This means that `visualStart` doesn't only not have the velocity offset that projectiles do, but also isn't rotated by the hardpoint offset anymore.

**Comment by DeBlister** [comment]
> The "start" Point is used to correctly place projectiles but it was incorrectly also used to place effects. (The velocity inclusion is a positional offset, and doesn't impact actual velocity velocity of anything. Purely positional fiddlings.)
> 
> I would not be so sure it's a workaround, I think applying "visualStart" in other places, where "start" is used now, would introduce the opposite bug in projectiles- appearing ahead of their parent ship instead of perfectly on top.

You're correct about projectiles experiencing the same issue if you remove the position offset based on velocity.  It seems that we can't have both without checking what kind of projectile is being launched.  Either that or something else is being done differently between effects and projectiles.

File: source/Hardpoint.cpp (hunk 1 above)
Comment (tehhowch) [reply source/Hardpoint.cpp:219]: The `0.5 * velocity` offset is explained more in BatchDrawList:
https://github.com/endless-sky/endless-sky/blob/f5f638c7c4a94b2b37d4c2006e0c50d4e8666c7f/source/BatchDrawList.cpp#L62-L66

File: source/Hardpoint.cpp [hunk 4]
```diff
@@ -218,16 +218,18 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
...
 
 	// Apply the aim and hardpoint offset.
 	aim += angle;
 	start += aim.Rotate(outfit->HardpointOffset());
+	visualStart += aim.Rotate(outfit->HardpointOffset());
```
Comment (tehhowch) [inline source/Hardpoint.cpp:226]: Why are we doing the same thing twice?

The only difference between `start` and `visualStart` is a constant offset (` - .5 * ship.Velocity()`), so we should only apply the correction when creating the visuals. Then the math is kept in sync without duplication of effort.

If it's most clear in terms of variable naming and providing context for the offset, first create what is currently termed as "visual start" but call it something like "point of origin," and then introduce the variable used for the projectile's initial drawn position with a suitable comment explaining the need for the particular offset.

File: source/Hardpoint.cpp [hunk 5]
```diff
@@ -218,16 +218,18 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
...
 
 	// Create a new projectile, originating from this hardpoint.
 	projectiles.emplace_back(ship, start, aim, outfit);
 
 	// Create any effects this weapon creates when it is fired.
-	CreateEffects(outfit->FireEffects(), start, ship.Velocity(), aim, visuals);
+	CreateEffects(outfit->FireEffects(), visualStart, ship.Velocity(), aim, visuals);
 
 	// Update the reload and burst counters, and expend ammunition if applicable.
 	Fire(ship, start, aim);
```
Comment (tehhowch) [inline source/Hardpoint.cpp:235]: This should probably use the point of origin as well, rather than the projectile draw position.

**Review by tehhowch — CHANGES_REQUESTED** [review]

File: source/Hardpoint.cpp [hunk 6]
```diff
@@ -218,16 +218,18 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
 	// offset of (.5 * velocity) and that velocity includes the velocity of the
 	// ship that fired them.
 	Point start = ship.Position() + aim.Rotate(point) - .5 * ship.Velocity();
+	Point visualStart = ship.Position() + aim.Rotate(point);
```
Comment (Amazinite) [inline source/Hardpoint.cpp:221]:
```suggestion
	Point start = ship.Position() + aim.Rotate(point);
```

File: source/Hardpoint.cpp (hunk 4 above)
Comment (Amazinite) [inline source/Hardpoint.cpp:226]:
```suggestion
```

File: source/Hardpoint.cpp [hunk 7]
```diff
@@ -218,16 +218,18 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
...
 	aim += angle;
 	start += aim.Rotate(outfit->HardpointOffset());
+	visualStart += aim.Rotate(outfit->HardpointOffset());
 
 	// Create a new projectile, originating from this hardpoint.
 	projectiles.emplace_back(ship, start, aim, outfit);
 
 	// Create any effects this weapon creates when it is fired.
-	CreateEffects(outfit->FireEffects(), start, ship.Velocity(), aim, visuals);
+	CreateEffects(outfit->FireEffects(), visualStart, ship.Velocity(), aim, visuals);
```
Comment (Amazinite) [inline source/Hardpoint.cpp:232]:
```suggestion
	CreateEffects(outfit->FireEffects(), start, ship.Velocity(), aim, visuals);
```

File: source/Hardpoint.cpp [hunk 8]
```diff
@@ -218,16 +218,18 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
...
 	Point start = ship.Position() + aim.Rotate(point) - .5 * ship.Velocity();
+	Point visualStart = ship.Position() + aim.Rotate(point);
 
 	// Apply the aim and hardpoint offset.
 	aim += angle;
 	start += aim.Rotate(outfit->HardpointOffset());
+	visualStart += aim.Rotate(outfit->HardpointOffset());
 
 	// Create a new projectile, originating from this hardpoint.
 	projectiles.emplace_back(ship, start, aim, outfit);
```
Comment (Amazinite) [inline source/Hardpoint.cpp:229]:
```suggestion
	// In order to get projectiles to start at the right position they are drawn
	// at an offset of (.5 * velocity).
	projectiles.emplace_back(ship, start - .5 * ship.Velocity(), aim, outfit);
```

File: source/Hardpoint.cpp [hunk 9]
```diff
@@ -218,16 +218,18 @@ void Hardpoint::Fire(Ship &ship, vector<Projectile> &projectiles, vector<Visual>
 	// offset of (.5 * velocity) and that velocity includes the velocity of the
 	// ship that fired them.
```
Comment (Amazinite) [inline source/Hardpoint.cpp:219]: GitHub won't let me push a suggestion high enough, but this comment should now be removed since I've suggested it be moved down.

**Comment by Ferociousfeind** [comment]
If I understand correctly, and apologies for the week-long wait... this would be the agreed-upon, better implementation? It is a little hard to follow the individual suggestions one line at a time, but I believe the consensus is "explicitly add the offset only to the projectiles that need them, and you can do that directly in the call to create them to skip making a new ancillary variable", and a point to BatchDrawList.cpp as an explanation as to why we do this.
image

If this is the consensus, I can push this, plus any further suggestion, to the branch, which I think would outdate and wipe out the suggested changes above, but I think it addresses them sufficiently...?

**Comment by Amazinite** [comment]
Looks good to me.

**Review by tehhowch — APPROVED** [review]
